rotkehlchen/chain/evm/decoding/pendle/decoder.py

```diff
@@ -1,122 +1,317 @@
+                event.notes = f'Return {event.amount} {token_1.symbol} to Pendle'
 
         return DEFAULT_DECODING_OUTPUT
 
+    def _decode_pendle_swap(self, context: DecoderContext) -> DecodingOutput:
+        if context.tx_log.topics[0] != SWAP_SINGLE_TOPIC:
+            return DEFAULT_DECODING_OUTPUT
+
+        # Determine swap type and retrieve received token details.
+        # Pendle only emits an event for the token being swapped,
+        # so we must check transaction logs to find the received token
+        token_out = (
+            self.evm_inquirer.native_token
+            if (addr := bytes_to_address(context.tx_log.topics[2])) == ZERO_ADDRESS
+            else self.base.get_or_create_evm_token(addr)
+        )
+        amount_out = asset_normalized_value(
+            amount=int.from_bytes(context.tx_log.data[:32]),
+            asset=token_out,
+        )
+        # no transaction was found that touches 1inch so we don't check for it(i.e swap_type=5).
+        if (swap_type := int.from_bytes(context.tx_log.topics[1])) not in (1, 2, 4):
+            return DEFAULT_DECODING_OUTPUT
+
+        for tx_log in context.all_logs:
+            if (  # odos
+                    swap_type == 2 and
+                    tx_log.topics[0] == ODOS_SWAP_TOPIC and
+                    bytes_to_address(tx_log.data[:32]) == PENDLE_SWAP_ADDRESS
+            ):
+                token_in = (
+                    self.evm_inquirer.native_token
+                    if (addr := bytes_to_address(tx_log.data[128:160])) == ZERO_ADDRESS
+                    else self.base.get_or_create_evm_token(addr)
+                )
+                amount_in = asset_normalized_value(
+                    amount=int.from_bytes(tx_log.data[96:128]),
+                    asset=token_in,
+                )
+                break
+
+            elif (  # kyberswap
+                    swap_type == 1 and
+                    tx_log.topics[0] == KYBERSWAP_SWAPPED_TOPIC and
+                    bytes_to_address(tx_log.data[96:128]) == PENDLE_ROUTER_ADDRESS
+            ):
+                token_in = (
+                    self.evm_inquirer.native_token
+                    if (addr := bytes_to_address(tx_log.data[64:96])) == ETH_SPECIAL_ADDRESS
+                    else self.base.get_or_create_evm_token(addr)
+                )
+                amount_in = asset_normalized_value(
+                    amount=int.from_bytes(tx_log.data[160:192]),
+                    asset=token_in,
+                )
+                break
+
+            elif (  # okx
+                swap_type == 4 and
+                tx_log.topics[0] == OKX_ORDER_RECORD_TOPIC
+            ):
+                token_in = (
+                    self.evm_inquirer.native_token
+                    if (addr := bytes_to_address(tx_log.data[32:64])) == ETH_SPECIAL_ADDRESS
+                    else self.base.get_or_create_evm_token(addr)
+                )
```

Comment:
I see that you do this often. Maybe you can make this a util func in `self.base`? I believe we might use it in other decoders